packages/types/src/codec/Enum.ts

```diff
@@ -2,94 +2,209 @@
 // This software may be modified and distributed under the terms
 // of the Apache-2.0 license. See the LICENSE file for details.
 
-import { isString, isU8a, u8aToHex, assert } from '@polkadot/util';
+import { Codec, Constructor } from '../types';
 
-import { Codec } from '../types';
+import { assert, hexToU8a, isHex, isNumber, isObject, isString, isU8a, isUndefined, u8aConcat, u8aToHex } from '@polkadot/util';
+
+import Null from '../primitive/Null';
 import Base from './Base';
 
-type EnumMap = {
-  [index: number]: string
+type EnumConstructor<T = Codec> = { new(value?: any, index?: number): T };
+
+type TypesDef = {
+  [name: string]: Constructor
+};
+
+type Decoded = {
+  index: number,
+  value: Codec
 };
 
-type EnumDef = {
-  [index: string]: number
-} | Array<string>;
+type Aliasses = {
+  [index: string]: string
+};
 
 /**
  * @name Enum
  * @description
- * A codec wrapper for an enum. Enums are encoded as a single byte, where the byte
- * is a zero-indexed value. This class allows you to retrieve the value either
- * by `toNumber()` exposing the actual raw index, or `toString()` returning a
- * string representation (as provided as part of the constructor)
+ * This implements an enum, that based on the value wraps a different type. It is effectively
+ * an extension to enum where the value type is determined by the actual index.
  */
 // TODO:
-//   - It would be great if this could actually wrap actual TS enums
-export default class Enum extends Base<number> implements Codec {
-  private _enum: EnumMap | Array<string>;
+//   - As per Enum, actually use TS enum
+//   - It should rather probably extend Enum instead of copying code
+export default class Enum extends Base<Codec> implements Codec {
```

Comment:
Fookit! The rename created issues here, now it looks like the actual file was adapted :(

Reply:
My first force-push to a PR to make this sane.